### ISSUE TITLE
Option to invoke API Gateway Endpoint instead of Lambda

### ISSUE THREAD

**Issue by ac360** [serverless/serverless#2210]
# This is a Feature Proposal

### Problem

When developing a REST API, using `sls invoke` only invokes the function.  This approach is useful, but it would be helpful to invoke the API Gateway Endpoint for a couple of reasons:

* `sls invoke -l` is great because you can instantly get logs.  But when I want to test an attached endpoint, I have to use CURL or Postman w/ `sls logs -t` which is a lot slower and it slows down my development process.
* I don't like using another tool to test my Lambda-backed endpoints.  I'd like the Framework to do it, since it has access to `event.json` and my http endpoint config already.  Retyping this elsewhere in another tool should not be required.  It slows down the development process.

### Solution

Allow passing in a flag to optionally call the REST API endpoint, instead of the function.  The flag should be something like `--endpoint`.

It's used like this:

```
sls invoke -f usersCreate -l -e // without specific endpoint
sls invoke -f usersCreate -l -e [method~path] // with specific endpoint, separate method and path via tilde: ~
```

It does this:

* If an endpoint is specified `sls invoke -f usersCreate -e post~users/create`, choose that endpoint.  
* If an endpoint is not specified, choose the first endpoint in the `events` array.
* Model an http request after the endpoint's configuration.
* Perform request
* Return response as if it were a normal Lambda invocation.  
* Return some extra information on the HTTP Request, like response time.

**Comment by pouledodue** [comment]
looks more and more like a needed swagger+sls integration
with tools like [swagger-tester](https://github.com/Trax-air/swagger-tester)

**Comment by dougmoscrop** [comment]
What about a separate command, `sls event {eventType}` -- `sls event http`, `sls event sns`, etc since a function can be subscribed to different events. It would actually be awesome to be able to have a event generator plugin that someone makes, and so on. I stumbled a bit when getting started on what AWS events for 'real' things looked like.

"invoke" staying.. well, "invoke".

**Comment by flomotlik** [comment]
Yup, totally would love to have this. I'm not sure yet about adding this to invoke because it might confuse invoke a bit and I think we'll need additional parameters like headers, ... for it. 

One option could be `serverless invoke http` or `serverless invoke endpoint` so its part of invoke, but its a separate command so not confusing. Then in the future we can also add other `serverless invoke ...` commands if we want to.

And as much as I like curl as a tool it really doesn't have the best usability and we should probably have that built-in with nice coloring, ...

What could be really interesting additionally is adding the ability to validate the response, e.g. you can define a file that has config for the expected Headers, status code, content, ... (although I think this should be implemented in a future iteration on this)

Options it needs to have:

* Header
* Method (from serverless.yml as well)
* Path (from serverless.yml)
* Body

From Method/Path we could probably also get the function specifically, but we might need them to add the `-f` config as well so we can load the logs (e.g. with variables it can be harder to be sure which function was called).

@serverless/vip

**Comment by svdgraaf** [comment]
'invoke http' makes sense, as it maps directly to the serverless.yml config. It would be nice if that is integrated into the framework (instead of using curl).

There already is a PR for (required) request parameters. Post params / models is a different discussion though (openapi/swagger). Perhaps that would be better solved by just fetching the open api definition from apig and feeding it to a swagger client of sorts. That would make api integration testing possible as well.

**Comment by dinvlad** [comment]
What about IAM-protected endpoints? We can only use those with an AWS-generated SDK, otherwise one needs to use a 3rd-party library to sign the custom requests (unfortunately, it doesn't seem like native AWS SDK provides this ability).

**Comment by erikerikson** [comment]
@dinvlad take a look at https://github.com/Nordstrom/artillery-plugin-aws-sigv4/blob/master/lib/aws-sigv4.js for how to sign with aws-sdk

**Comment by dinvlad** [comment]
Thanks, I was aware of the procedure and just pointing out here that support for calling the endpoints directly should also cover signed requests. Also, it'd be ideal to use a well-tested and supported library (preferably official if it existed).

**Comment by erikerikson** [comment]
The aws-sdk doesn't seem well tested, supported, or official.  Interesting.  Agreed that official docs are in order but the referenced code is what the sdk uses internally and there are official AWS docs that promote this usage so while you are correct, for anyone else reviewing this conversation, I would suggest that the risk profile is incredibly low.

**Comment by dinvlad** [comment]
Internal APIs are always prone to changes that may eventually break things. I appreciate that your suggestion probably works well at the moment, but personally I'd favor something akin to mhart/aws4 because of its reliance on the actual signing algorithm and test coverage. Thanks

**Comment by dinvlad** [comment]
In addition, I have tried using that API via the current version of the SDK, and there doesn't appear to be a V4 object under aws.Signers. Please correct me if I'm wrong. Thanks

**Comment by dinvlad** [comment]
FWIW, here's the official response from the AWS team: https://github.com/aws/aws-sdk-js/issues/1247#issuecomment-264285589 - they don't recommend using that API at the moment.

**Comment by erikerikson** [comment]
Thanks for that link and the corresponding discussion @dinvlad, I agree with your position that the generic capability being desirable.  We definitely had some pain points on that rough corner

**Comment by erikerikson** [comment]
The downside is reduced local management of invalid requests and increased cost as a result
